### ISSUE TITLE
Depth Based Block Replacement

### ISSUE THREAD

**Issue by ph3rr3t** [pacas00/ModdersGearbox#7]
So many neat things I could do with a feature like this.
Sample code below, might be a few other parameters that could be added that I've not thought of;

```
[GenerationReplacementEntry]
	[TargetKey]DeepRock[/TargetKey]
	[ReplacementKey]NewRockOrStone[/ReplacementKey]
	[ReplacementValue]0[/ReplacementValue]  <!-- Optional, Defaults to default value, optionally, put a ValueEntry key instead of a number and have the code lookup the value -->
	[ReplacementChance]25[/ReplacementChance] <!-- Optional, 0-100 -->
	[MinDepth]-1000[/MinDepth]
	[MaxDepth]-2000[/MaxDepth]
[/GenerationReplacementEntry]
```

**Comment by pacas00** [comment]
@ph3rr3t Do you want to be able to set a chance of this being replaced or anything?

**Comment by pacas00** [comment]
Alright, I've updated the example entry with what i feel is a cleaner set of names.

Quite likely, this will get run after the built in mod ore gen code.

**Comment by ph3rr3t** [comment]
What is the purpose of the [ReplacementValue] line?

**Comment by pacas00** [comment]
To allow a value to be set on the replacement block,

For paintable or colourised blocks, it's the colour.
For anything with a ValueEntry, then a specific key to use. (or number if your using vanilla cubes)

**Comment by ph3rr3t** [comment]
Would that work with an ore block's value too?

**Comment by pacas00** [comment]
Yes, but there is a terrain generator for mod ores already in the game

**Comment by pacas00** [comment]
Though in saying so, if you wanted an ore, but random low chance single blocks and not veins, this would work for that.

**Comment by ph3rr3t** [comment]
You're the coder so you get to pick the tags, but I had a thought;

[BlockReplaceEntry]
-[OldKey]DeepRock[/OldKey]
-[NewKey]CustomKey[/NewKey]
-[NewValue]0[/NewValue] 
 ----- 
 -[NewChance]25[/NewChance] 
 -[MinDepth]-1000[/MinDepth]
 -[MaxDepth]-2000[/MaxDepth]
[/BlockReplaceEntry]

**Comment by ph3rr3t** [comment]
> 
> 
> So many neat things I could do with a feature like this.
> Sample code below, might be a few other parameters that could be added that I've not thought of;
> 
> ```
> [GenerationReplacementEntry]
> 	[TargetKey]DeepRock[/TargetKey]
> 	[ReplacementKey]NewRockOrStone[/ReplacementKey]
> 	[ReplacementValue]0[/ReplacementValue]  
> 	[ReplacementChance]25[/ReplacementChance] 
> 	[MinDepth]-1000[/MinDepth]
> 	[MaxDepth]-2000[/MaxDepth]
> [/GenerationReplacementEntry]
> ```

**Comment by pacas00** [comment]
Please don't close this, Unless you no longer want it. Otherwise i'll forget it.

**Comment by ph3rr3t** [comment]
If I closed it then I don't know how I did it.  Thought you had done so.  As for the tags, lets just go with the ones you put in OP.

**Comment by pacas00** [comment]
Bumping this forward to v1.2, Everything below here is notes so i know where to write the implementation of this.

Todo

- [ ] Write data class
- [ ] write loading code
- [ ] Hook Postfix on GenerateCubes in SurvivalUnderGroundGenerator and SurvivalUndergroundCaveGenerator
- [ ] Loop though replacements and maCubes till we get a hit, roll dice and replace
- [ ] Handle values
- [ ] Test

PerlinCache cache stores the cubes?
arrays are [16,16,16]
maCubes and maCubeValues

**Comment by pacas00** [comment]
So, basically ore gen? Why not just use ore gen for that?

**Comment by ph3rr3t** [comment]
Kinda as biome layers.  Currently flatland only affects the surface.  This would allow us to modify everything below ground as well.

**Comment by pacas00** [comment]
#19
